types/trace.go

```diff
@@ -0,0 +1,132 @@
+package types
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+)
+
+// TraceSession model.
+type TraceSession struct {
+	ID         string    `json:"id" yaml:"id"`
+	PipelineID string    `json:"pipelineID" yaml:"pipelineID"`
+	Plugins    []string  `json:"plugins" yaml:"plugins"`
+	Lifespan   Duration  `json:"lifespan" yaml:"lifespan"`
+	CreatedAt  time.Time `json:"createdAt" yaml:"createdAt"`
+	UpdatedAt  time.Time `json:"updatedAt" yaml:"updatedAt"`
+}
+
+// Active tells whether a session is still within its lifespan.
+func (ts TraceSession) Active() bool {
+	return ts.CreatedAt.Add(time.Duration(ts.Lifespan)).After(time.Now())
+}
+
+// CreateTraceSession request payload for creating a new trace session.
+type CreateTraceSession struct {
+	Plugins  []string `json:"plugins"`
+	Lifespan Duration `json:"lifespan"`
+}
+
+// TraceSessionsParams request payload for querying trace sessions.
+type TraceSessionsParams struct {
+	Last   *uint64
+	Before *string
+}
+
+// TraceSessions paginated list.
+type TraceSessions struct {
+	Items     []TraceSession `json:"items"`
+	EndCursor *string        `json:"endCursor"`
+}
+
+// UpdateTraceSession request payload for updating a trace session.
+type UpdateTraceSession struct {
+	Plugins  *[]string `json:"plugins"`
+	Lifespan *Duration `json:"lifespan"`
+}
+
+// TraceRecord model.
+type TraceRecord struct {
+	ID        string    `json:"id" yaml:"id"`
+	SessionID string    `json:"sessionID" yaml:"sessionID"`
+	CreatedAt time.Time `json:"createdAt" yaml:"createdAt"`
+
+	// fluent-bit data from here on.
+	Kind           TraceRecordKind `json:"type" yaml:"type"`
```

Comment:
The actual type parameter is an integer in the trace record itself. Is this handled via a Unmarshal or something similar?

Reply:
I didn't knew. I'll change it to be a integer 👍

Reply:
What is the order?

Reply:
@pwhelan what is the order?

Reply:
The order, documented here https://github.com/fluent/fluent-bit/blob/6d4cc77236210d6989281c24f4664c9489072d80/include/fluent-bit/flb_trace_chunk.h#L9, is:

- INPUT: 1
- FILTER: 2
- PRE_OUTPUT: 3
- OUTPUT: 4

I do not foresee any other event types. The only type of event we might need to add would be for the different filter results, ie: dropped, modified, ignored, unmodified, etc... but I would add that as a whole different property so no worries there.

Reply:
Nice. Thanks.

Reply:
Done.